### ISSUE TITLE
kpatch load look leads to deadlock in stop_machine context

### ISSUE THREAD

**Issue by xiejingf** [dynup/kpatch#1077]
I registered a look hook using KPATCH_LOAD_HOOK(foo) and encounter the scenario below which leads to system deadlock

stop_machine -> foo,   in the foo function , acquire a mutex failed,   then schedule,  the new picked up task happens to send ipi across other CPUs and busy wait other CPUs to respond, which is impossible, because the other CPUs are stopped by stop_machine call,  so the system resulted in a deadlock state.

My suggestion is move the kpatch load/unload hook outside the stop_machine context, like module_init in kernel live patch.

My patch is based on v0.3.2

```
diff --git a/kmod/core/core.c b/kmod/core/core.c
index 77c7290..9ffa57b 100644
--- a/kmod/core/core.c
+++ b/kmod/core/core.c
@@ -413,13 +413,6 @@ static int kpatch_apply_patch(void *data)
 		return -EBUSY;
 	}
 
-	/* run any user-defined load hooks */
-	list_for_each_entry(object, &kpmod->objects, list) {
-		if (!kpatch_object_linked(object))
-			continue;
-		list_for_each_entry(hook, &object->hooks_load, list)
-			(*hook->hook)();
-	}
 
	return 0;
@@ -450,13 +443,6 @@ static int kpatch_remove_patch(void *data)
 		hash_del_rcu(&func->node);
 	} while_for_each_linked_func();
 
-	/* run any user-defined unload hooks */
-	list_for_each_entry(object, &kpmod->objects, list) {
-		if (!kpatch_object_linked(object))
-			continue;
-		list_for_each_entry(hook, &object->hooks_unload, list)
-			(*hook->hook)();
-	}
 
 	return 0;
 }
@@ -923,6 +909,7 @@ int kpatch_register(struct kpatch_module *kpmod, bool replace)
 	int ret, i, force = 0;
 	struct kpatch_object *object, *object_err = NULL;
 	struct kpatch_func *func;
+	struct kpatch_hook *hook;
 
 	if (!kpmod->mod || list_empty(&kpmod->objects))
 		return -EINVAL;
@@ -971,6 +958,16 @@ int kpatch_register(struct kpatch_module *kpmod, bool replace)
 
 	kpatch_state_updating();
 
+
+	printk("running user defined load hooks");
+	/* run any user-defined load hooks */
+	list_for_each_entry(object, &kpmod->objects, list) {
+		if (!kpatch_object_linked(object))
+			continue;
+		list_for_each_entry(hook, &object->hooks_load, list)
+			(*hook->hook)();
+	}
+
 	/*
 	 * Idle the CPUs, verify activeness safety, and atomically make the new
 	 * functions visible to the ftrace handler.
@@ -1076,6 +1073,7 @@ int kpatch_unregister(struct kpatch_module *kpmod)
 {
 	struct kpatch_object *object;
 	struct kpatch_func *func;
+	struct kpatch_hook *hook;
 	int ret, force = 0;
 
 	down(&kpatch_mutex);
@@ -1098,6 +1096,14 @@ int kpatch_unregister(struct kpatch_module *kpmod)
 
	 ret = stop_machine(kpatch_remove_patch, kpmod, NULL);
 
+	/* run any user-defined unload hooks */
+	list_for_each_entry(object, &kpmod->objects, list) {
+		if (!kpatch_object_linked(object))
+			continue;
+		list_for_each_entry(hook, &object->hooks_unload, list)
+			(*hook->hook)();
+	}
+
 	/* NMI handlers can return to normal now */
 	kpatch_state_idle();
```

**Comment by jpoimboe** [comment]
kpatch.ko is deprecated in favor of CONFIG_LIVEPATCH, I will leave this to @euspectre to comment :-)

**Comment by xiejingf** [comment]
> kpatch.ko is deprecated in favor of CONFIG_LIVEPATCH, I will leave this to @euspectre to comment :-)

we use kernel 3.10 on mass product system, which does not support kernel live patch :-)

**Comment by jpoimboe** [comment]
Understood.  Feel free to send a pull request.

**Comment by euspectre** [comment]
Yes, running load/unload hooks under stop_machine() in kpatch.ko makes them way less useful than they could be.

So, you need to call a function that may sleep/schedule from the hook `foo`, right?

Looking forward to your patch(es), it is better to start from there.

**Comment by xiejingf** [comment]
> So, you need to call a function that may sleep/schedule from the hook `foo`, right?
> 
> Yes, running load/unload hooks under stop_machine() in kpatch.ko makes them way less useful than they could be.
> 
> I suspect that if we simply move the hook outside of stop_machine(), we might get into a situation when the patched code starts running concurrently with the hook function. This could be disastrous in certain cases.
> 
> Looking forward to your patch(es) anyway, it is better to start from there.
> 
> Or, perhaps, if you could provide more info about the live patch you are working on, we could find a workaround for this particular case, leaving the code of kpatch.ko as it is.

**Comment by xiejingf** [comment]
@euspectre , many user use kpatch as a tool to build hot fixes, so it is hard to let them work around their own codes one by one, I think it is better for kpatch to fix this

**Comment by euspectre** [comment]
My point is, the new (patched) functions may start right after `stop_machine(kpatch_apply_patch, &args, NULL)` returns. So, if we run the hooks at that stage rather than under `stop_machine()`, the hooks may end up running in parallel with the patched functions, which may be unsafe.

Even if some hook functions could safely run concurrently with the patched code, some may require that they complete before any patched code runs.

For example, some of the KPatch-based patches we deployed in our production systems used the load hooks to prepared system state required for the patched code. It was needed that the state was consistent before the pacthed code had started.

>  I think it is better for kpatch to fix this

I agree, but if we find that this is not possible, other solutions should be prepared.

For example, if the code you would like to call from the load hook can safely run concurrently with the patched code, you can move it into a workqueue item and make the patched code wait for it, where appropriate. The load hook will now just create that wq item and queue it. IIRC, this can be done from stop_machine().

**Comment by joe-lawrence** [comment]
> For example, some of the KPatch-based patches we deployed in our production systems used the load hooks to prepare system state required for the patched code. It was needed that the state was consistent before the patched code had started.

Hi Evgenii -- any chance you can share code (or at least generalize) to demonstrate this callback use-case?  I only ask since the upstream consistency model callbacks do not rely upon stop_machine.  We've always held stop_machine in our back pocket in case we ever did need some kind of machine consistency, but I don't think I've ever actually used it for kpatch or upstream.  This might make an interesting addition to the upstream callback documentation.

**Comment by euspectre** [comment]
> any chance you can share code (or at least generalize) to demonstrate this callback use-case?

I don't remember the details now but, IIRC, the load hook initialized some data structure which was then used by the patched code. It could be possible to do lazy initialization in the patched code, use some synchronization to prevent racy init, and avoid using the hooks altogether, but it would be less convenient.

What we actually needed there, was a pre-patch callback. However, it was before you had added support for such callbacks.

> I only ask since the upstream consistency model callbacks do not rely upon stop_machine.

Yes, but, stop_machine is not needed there, I suppose.

In livepatch, pre-patch callbacks are guaranteed to complete before any patched code starts, right? This guarantee would be enough in our case.

As for post-patch callbacks, we have yet to encounter a situation in our patches, where they would be absolutely necessary. If they can run concurrently with the patched code, it seems OK.

**Comment by joe-lawrence** [comment]
> > any chance you can share code (or at least generalize) to demonstrate this callback use-case?
> 
> I don't remember the details now but, IIRC, the load hook initialized some data structure which was then used by the patched code. It could be possible to do lazy initialization in the patched code, use some synchronization to prevent racy init, and avoid using the hooks altogether, but it would be less convenient.
> 
> What we actually needed there, was a pre-patch callback. However, it was before you had added support for such callbacks.
> 
> > I only ask since the upstream consistency model callbacks do not rely upon stop_machine.
> 
> Yes, but, stop_machine is not needed there, I suppose.
> 
> In livepatch, pre-patch callbacks are guaranteed to complete before any patched code starts, right? This guarantee would be enough in our case.
> 
> As for post-patch callbacks, we have yet to encounter a situation in our patches, where they would be absolutely necessary. If they can run concurrently with the patched code, it seems OK.

**Comment by xiejingf** [comment]
> My point is, the new (patched) functions may start right after stop_machine(kpatch_apply_patch, &args, NULL) returns. So, if we run the hooks at that stage rather than under stop_machine(), the hooks may end up running in parallel with the patched functions, which may be unsafe.

@euspectre ，we run the load hook before call `stop_machine`, so it is guaranteed that load hook completes before patched code start, do you agree ?

**Comment by euspectre** [comment]
The current implementation of kpatch core has 4 kinds of callbacks: pre-patch, post-patch, pre-unpatch and post-unpatch.

I think it would be OK to run pre-patch callbacks before `stop_machine(kpatch_apply_patch, <...>)` and post-unpatch callbacks - after `stop_machine(kpatch_remove_patch, <...>)`.

As for post-patch & pre-unpatch hooks, I'd leave them under stop_machine(), just in case, someone needs stronger guarantees there.

**Comment by euspectre** [comment]
First, I thought you were going to move all hooks outside of stop_machine(), both pre- and post-. That could be problematic. I missed that you use kpatch 0.3 where there were only 2 hooks, no pre-/post-.

Now that the hooks have been split into pre- & post- parts, I think, it is OK to relax the requirements on pre-patch and post-unpatch.

Looking forward to your pull request.

**Comment by xiejingf** [comment]
@euspectre 
>  I missed that you use kpatch 0.3 where there were only 2 hooks, no pre-/post-

Yes，we still use kpatch 0.3 in our production system, so my suggestions are based on it

> For example, if the code you would like to call from the load hook can safely run concurrently with the patched code, you can move it into a workqueue item and make the patched code wait for it, where appropriate. The load hook will now just create that wq item and queue it. IIRC, this can be done from stop_machine()

schedule a work_struct in stop machine ,then in kpatch_ftrace_handler  wait for the work to complete, it is an idea, I will try.

> Looking forward to seeing your pull request.

So, my pull request need to based on the latest kpatch version(v0.9) , right?

**Comment by xiejingf** [comment]
> For example, if the code you would like to call from the load hook can safely run concurrently with the patched code, you can move it into a workqueue item and make the patched code wait for it, where appropriate. The load hook will now just create that wq item and queue it. IIRC, this can be done from stop_machine()

It could be problematic, for example, if the patched func runs in interrupt context, which sleeping is forbidden !

**Comment by euspectre** [comment]
> So, my pull request need to based on the latest kpatch version(v0.9) , right?

Yes, it should be based on the 'master' branch in this git repo.

> It could be problematic, for example, if the patched func runs in interrupt context, which sleeping is forbidden !

Yes, that is why I wrote "where appropriate". If you schedule a work item, you should decide where in the patched code it is appropriate to wait for it.

**Comment by joe-lawrence** [comment]
Fixed by PR #1134 ("deal with the deadlock may occur in stop_machine context")
